applications/tari_conduit/src/main.rs

```diff
@@ -0,0 +1,630 @@
+    assert!(count >= 3);
+    assert!(count <= 0x10000000);
+
+    let mut pow: usize = 2;
+    while pow < count {
+        pow <<= 1;
+    }
+    pow >> 1
+}
+
+// Tree hash algorithm in monero
+#[allow(clippy::needless_range_loop)]
+pub fn tree_hash(hashes: Vec<Hash>) -> Vec<u8> {
+    assert!(!hashes.is_empty());
+    match hashes.len() {
+        1 => hashes[0].0.to_vec(),
+        2 => {
+            let mut buf: [u8; 64] = [0; 64];
+            buf[..32].copy_from_slice(&hashes[0].0.to_vec());
+            buf[32..].copy_from_slice(&hashes[1].0.to_vec());
+            cn_fast_hash(&buf)
+        },
+        _ => {
+            let mut cnt = tree_hash_cnt(hashes.len());
+            let mut buf: Vec<u8> = Vec::with_capacity(cnt * 32);
+
+            for i in 0..(2 * cnt - hashes.len()) {
+                for val in &hashes[i].0.to_vec() {
+                    buf.push(val.to_owned());
+                }
+            }
+
+            for _i in (2 * cnt - hashes.len())..(cnt * 32) {
+                buf.push(0);
+            }
+
+            let mut i: usize = 2 * cnt - hashes.len();
+            for j in (2 * cnt - hashes.len())..cnt {
+                let mut tmp: [u8; 64] = [0; 64];
+                tmp[..32].copy_from_slice(&hashes[i].0.to_vec());
+                tmp[32..].copy_from_slice(&hashes[i + 1].0.to_vec());
+                let tmp = cn_fast_hash(&tmp);
+                buf[(j * 32)..((j + 1) * 32)].copy_from_slice(tmp.as_slice());
+                i += 2;
+            }
+            assert_eq!(i, hashes.len());
+
+            while cnt > 2 {
+                cnt >>= 1;
+                let mut i = 0;
+                for j in (0..(cnt * 32)).step_by(32) {
+                    let tmp = cn_fast_hash(&buf[i..(i + 64)]);
+                    buf[j..(j + 32)].copy_from_slice(tmp.as_slice());
+                    i += 64;
+                }
+            }
+
+            cn_fast_hash(&buf[..64])
+        },
+    }
+}
+
+fn get_monero_data(data: &[u8], seed: String) -> Result<MoneroData, Error> {
+    // TODO: Params can be an array
+    let parsed = json::parse(&stringify_request(data)).unwrap();
+    let s = format!("{}", parsed["params"][0].clone());
+    let hex = hex::decode(s).unwrap();
+    let block = deserialize::<Block>(&hex).unwrap();
+    let mut hashes = block.clone().tx_hashes;
+    hashes.push(block.miner_tx.hash());
+    let root = tree_hash(hashes);
+    let mut proof = block.clone().tx_hashes;
+    proof.push(block.miner_tx.hash());
+    return Ok(MoneroData {
+        header: block.header.clone(),
+        key: seed,
+        count: (block.tx_hashes.len() as u16) + 1,
+        transaction_root: from_slice(root.as_slice()),
+        transaction_hashes: from_hashes(&proof),
+        coinbase_tx: block.miner_tx,
+    });
+}
+
+fn get_seed_hash(data: &[u8]) -> String {
+    let parsed = json::parse(&stringify_request(data)).unwrap();
+    let seed = &parsed["result"]["seed_hash"].to_string();
+    seed.to_owned()
+}
+
```

Comment:
Again, all of this is duplicity code, why dont we import this from tari::core.

Reply:
tari::core received the `tree_hash` functions from the conduit, will import from there.